File: nossi/3.Hash Table/Two-sum.py
# https://leetcode.com/problems/two-sum/

# dictionary를 쓸 수 있는 상황을 판별하는 게 굉장히 중요하다
# 시간복잡도를 엄청 낮출 수 있는 그런 자료구조이기 때문에 언제 어떻게 사용하는지

nums = [4,1,9,7,5,3,16]
target = 14

def two_sum(nums, target):
    memo = {}
    for v in nums:
        needed_number = target - v
        if needed_number in memo:
            return True
        memo[v] = 1
    return False

print(two_sum(nums, target))


def two_sum2(nums, target):
    memo = {}
    for i, v in enumerate(nums):
        if target - v in memo:
            return [memo[target - v], i]
        memo[v] = i
    return False

print(two_sum2(nums, target))

File: nossi/3.Hash Table/Two-sum.js
// https://leetcode.com/problems/two-sum/solutions/1606307/javascript-hash-table/?orderBy=most_votes&languageTags=javascript&topicTags=hash-table
// Hash
// Key-value쌍으로 데이터를 빠르게 찾아보세요.
// return indices of the two numbers

const nums = [4, 1, 9, 7, 5, 3, 16];
const target = 14;

function twoSum(nums, target) {
  const memo = {};
  for (let i = 0; i < nums.length; i++) {
    // memo[nums[i]] = i; 먼저 넣어주고 시작하면 자기 자신도 검사하므로 안됨 ex) target이 14일때 7이 들어오면 방금 막 넣어준 자기자신7과 일치해버림
    if (target - nums[i] in memo) {
      return [memo[target - nums[i]], i];
    }
    memo[nums[i]] = i;
  }
  return false;
}

console.log(twoSum(nums, target));
